include/mixer.h

```diff
@@ -480,4 +481,63 @@ void MIXER_SetReverbPreset(const ReverbPreset new_preset);
 ChorusPreset MIXER_GetChorusPreset();
 void MIXER_SetChorusPreset(const ChorusPreset new_preset);
 
+// Generic callback used for audio devices which generate audio on the main thread
+// These devices produce audio on the main thread and consume on the mixer thread
+// This callback is the consumer part
+template <class DeviceType, class AudioType, bool stereo, bool signeddata, bool nativeorder>
+inline void MIXER_PullFromQueueCallback(const int frames_requested, DeviceType* device)
+{
+	// Currently only handles mono sound (output_queue is a primitive type and frames == samples)
+	// Special case for AudioType == AudioFrame (stereo floating-point sound)
+	static_assert((!stereo) || std::is_same<AudioType, AudioFrame>::value);
+
+	// AudioFrame type is always stereo
+	static_assert(stereo || (!std::is_same<AudioType, AudioFrame>::value));
+
+	size_t clamped_frames = check_cast<size_t>(frames_requested);
+	if (MIXER_InFastForwardMode()) {
+		// Special case, normally only hit when using the fast-forward hotkey (Alt + F12)
+		// We need a very large buffer to compensate or it results in static
+		device->output_queue.Resize(iceil(device->channel->GetFramesPerBlock() * 100.0f));
+
+		// We must not block in fast-forward mode or the timings used by the mixer get confused
+		// Clamp and fill with silence if needed
+		clamped_frames = std::min(clamped_frames, device->output_queue.Size());
+	} else {
+		// Normal case, resize the queue to ensure we don't have high latency
+		// Resize is a fast operation, only setting a variable for max capacity
+		// It does not drop frames or append zeros to the end of the underlying data structure
+
+		// Size to 2x blocksize. The mixer callback will request 1x blocksize.
+		// This provides a good size to avoid over-runs and stalls.
+		device->output_queue.Resize(iceil(device->channel->GetFramesPerBlock() * 2.0f));
+	}
+	int frames_recieved = 0;
+	if (clamped_frames > 0) {
+		std::vector<AudioType> to_mix = {};
+		device->output_queue.BulkDequeue(to_mix, clamped_frames);
+		frames_recieved = check_cast<int>(to_mix.size());
+		if (frames_recieved > 0) {
+			// One of the GCC CI builds throws a duplicated branch warning
+			// Clang apparently doesn't have this warning because it throws an "unknown warning" warning in the pragma
+			#ifndef __clang__
+			#pragma GCC diagnostic push
+			#pragma GCC diagnostic ignored "-Wduplicated-branches"
+			#endif
+			if (std::is_same<AudioType, AudioFrame>::value) {
+				// AudioFrame has the same memory layout as 2x floats
+				device->channel->template AddSamples<float, stereo, signeddata, nativeorder>(frames_recieved, reinterpret_cast<float*>(to_mix.data()));
+			} else {
+				device->channel->template AddSamples<AudioType, stereo, signeddata, nativeorder>(frames_recieved, to_mix.data());
+			}
+			#ifndef __clang__
+			#pragma GCC diagnostic pop
+			#endif
```

Comment:
Eeeeek. This is really ugly to drop this in a middle of a function. I'd much rather surround the whole method with the pragma, or if you're hellbent on scoping it only these few lines, maybe extract it to an inline function and surround only that?

Reply:
I just like that this highlights exactly where the warning is.  Real fix is someone who knows templates better than me to clean this up so GCC doesn't warn :smile: